scripts/helper_scripts/unipept-database-rs/src/dat_parser/entry.rs

```diff
@@ -0,0 +1,532 @@
+use crate::uniprot::UniprotType;
+use anyhow::Context;
+use std::collections::HashSet;
+
+// Constants to aid in parsing
+const COMMON_PREFIX_LEN: usize = "ID   ".len();
+
+const ORGANISM_RECOMMENDED_NAME_PREFIX_LEN: usize = "RecName: Full=".len();
+const ORGANISM_RECOMMENDED_NAME_EC_PREFIX_LEN: usize = "EC=".len();
+const ORGANISM_TAXON_ID_PREFIX_LEN: usize = "OX   NCBI_TaxID=".len();
+const VERSION_STRING_FULL_PREFIX_LEN: usize = "DT   08-NOV-2023, entry version ".len();
+
+/// The minimal data we want from an entry out of the UniProtKB datasets
+pub struct UniProtDATEntry {
+    accession_number: String,
+    name: String,
+    sequence: String,
+    version: String,
+    ec_references: Vec<String>,
+    go_references: Vec<String>,
+    ip_references: Vec<String>,
+    taxon_id: String,
+}
+
+impl UniProtDATEntry {
+    /// Parse an entry out of the lines of a DAT file
+    pub fn from_lines(data: &mut Vec<String>) -> anyhow::Result<Self> {
+        let mut current_index: usize;
+
+        let mut version = String::new();
+        let mut name = String::new();
+        let mut ec_references = Vec::<String>::new();
+        let mut go_references = Vec::<String>::new();
+        let mut ip_references = Vec::<String>::new();
+        let mut taxon_id = String::new();
+        let mut sequence = String::new();
+
+        let accession_number = parse_ac_number(data).context("Error parsing accession number")?;
+        current_index = parse_version(data, &mut version);
+        current_index = parse_name_and_ec(data, current_index, &mut ec_references, &mut name);
+        current_index = parse_taxon_id(data, current_index, &mut taxon_id);
+        current_index =
+            parse_db_references(data, current_index, &mut go_references, &mut ip_references);
+        parse_sequence(data, current_index, &mut sequence);
```

Comment:
It feels a bit awkward to first make the resulting strings, and then pass them mutably to these `parse_...` functions. This file also feels quite large and complex.

Thinking out loud: would it be doable to make a separate `EntryParser` to have a flow like this:

```rust
let parser = EntryParser::new(data);
OK(Self {
	accession_number = parser.ac_number()?,
    version = parser.version()?,
    name = parser.name()?,
    ...
})
```

The `EntryParser` would then keep track of the `current_index` and other fields necessary to find the correct data. However, looking at the code of these functions, it seems quite complex to make this work.

Reply:
I don't think it is. Note that some of these functions parse multiple values at once (e.g. the name and ec numbers, the GO and InterPro references, etc.) because they are grouped together in the input file. Unless this is possible in Rust:
```rs
Self {
    (name, ec_references) = parser.name_and_ec_references()
}
```

I guess this could be solved by storing the extra ones inside of the `EntryParser` when the parsing function is called, and then simply retrieving them later, but this means you should **always** access certain fields in the same order, which I'm not necessarily a fan of. It feels quite error-prone to me, unless you feel like it'll be better than the current approach?

Reply:
I personally prefer avoiding passing `&mut` references if possible, but if that does not work with the current scheme that is alright. There is no need to change it if you don't think it would improve the readability of the code.

Reply:
It might improve readability, I agree with you the current setup is not ideal. I just think it's not possible due to the nature of the input file, causing us to parse multiple values at once. If we strictly require things to be called in a set order (which you could argue is the case now as well) then it could be possible. I just wanted to discuss if - taking this constraint into account - this is a better solution or not.

Reply:
I think the only "problem" I have with the current code, is passing the mutable strings. Just passing the `current_index` as mutual would feel better in my opinion:

```rust
        let mut current_index: usize;

        let accession_number = parse_ac_number(data).context("Error parsing accession number")?;
		let version = parse_version(data, &mut current_index);
        let (name, ec_references) = parse_name_and_ec(data, &mut current_index);
        let taxon_id = parse_taxon_id(data, &mut current_index);
        let (ec_references, go_references) = parse_db_references(data, &mut current_index);
        let sequence = parse_sequence(data, &mut current_index);
```

What do you think?

Reply:
I can get behind that, yeah. Definitely see how returning the strings ends up being cleaner than returning the index.